src/analysis/processing/qgsalgorithmdetectdatasetchanges.cpp

```diff
@@ -0,0 +1,432 @@
+                      "that these features have a unique set of attributes selected for comparison. If this condition is not met, warnings will be "
+                      "raised and the resultant outputs may be misleading.\n\n"
+                      "The algorithm outputs three layers, one containing all features which are considered to be unchanged between the revisions, "
+                      "one containing features deleted from the original layer which are not present in the revised layer, and one containing features"
+                      "add to the revised layer which are not present in the original layer." );
+}
+
+QString QgsDetectVectorChangesAlgorithm::shortDescription() const
+{
+  return QObject::tr( "Calculates features which are unchanged, added or deleted between two dataset versions." );
+}
+
+QgsDetectVectorChangesAlgorithm *QgsDetectVectorChangesAlgorithm::createInstance() const
+{
+  return new QgsDetectVectorChangesAlgorithm();
+}
+
+bool QgsDetectVectorChangesAlgorithm::prepareAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  mOriginal.reset( parameterAsSource( parameters, QStringLiteral( "ORIGINAL" ), context ) );
+  if ( !mOriginal )
+    throw QgsProcessingException( invalidSourceError( parameters, QStringLiteral( "ORIGINAL" ) ) );
+
+  mRevised.reset( parameterAsSource( parameters, QStringLiteral( "REVISED" ), context ) );
+  if ( !mRevised )
+    throw QgsProcessingException( invalidSourceError( parameters, QStringLiteral( "REVISED" ) ) );
+
+  mMatchType = static_cast< GeometryMatchType >( parameterAsEnum( parameters, QStringLiteral( "MATCH_TYPE" ), context ) );
+
+  switch ( mMatchType )
+  {
+    case Exact:
+      if ( mOriginal->wkbType() != mRevised->wkbType() )
+        throw QgsProcessingException( QObject::tr( "Geometry type of revised layer (%1) does not match the original layer (%2). Consider using the \"Tolerant Match\" option instead." ).arg( QgsWkbTypes::displayString( mRevised->wkbType() ),
+                                      QgsWkbTypes::displayString( mOriginal->wkbType() ) ) );
+      break;
+
+    case Topological:
+      if ( QgsWkbTypes::geometryType( mOriginal->wkbType() ) != QgsWkbTypes::geometryType( mRevised->wkbType() ) )
+        throw QgsProcessingException( QObject::tr( "Geometry type of revised layer (%1) does not match the original layer (%2)" ).arg( QgsWkbTypes::geometryDisplayString( QgsWkbTypes::geometryType( mRevised->wkbType() ) ),
+                                      QgsWkbTypes::geometryDisplayString( QgsWkbTypes::geometryType( mOriginal->wkbType() ) ) ) );
+      break;
+
+  }
+
+  if ( mOriginal->sourceCrs() != mRevised->sourceCrs() )
+    feedback->reportError( QObject::tr( "CRS for revised layer (%1) does not match the original layer (%2) - reprojection accuracy may affect geometry matching" ).arg( mOriginal->sourceCrs().userFriendlyIdentifier(),
```

Comment:
That reminds me we should consider a reportWarning function that would, in terms of semantic priority, sit in between exciting pushInfo and reportError.

Reply:
Yeah, I agree